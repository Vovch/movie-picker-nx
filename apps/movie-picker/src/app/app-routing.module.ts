import { NgModule } from '@angular/core';
import { RouterModule, Routes } from '@angular/router';
import {MovieListPageComponent} from './movie-list-page/movie-list-page.component';

const routes: Routes = [
  {path: '', component: MovieListPageComponent}
];

@NgModule({
  imports: [RouterModule.forRoot(routes)],
  exports: [RouterModule]
})
export class AppRoutingModule { }
