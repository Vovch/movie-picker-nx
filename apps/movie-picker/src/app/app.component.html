<movie-picker-header></movie-picker-header>

<section class="router-outlet">
    <router-outlet></router-outlet>
    <movie-picker-authentication-modal></movie-picker-authentication-modal>
    <movie-picker-error></movie-picker-error>
</section>
