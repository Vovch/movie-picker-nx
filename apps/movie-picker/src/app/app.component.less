@import "variables";

.router-outlet {
  padding: @default-offset;
}
