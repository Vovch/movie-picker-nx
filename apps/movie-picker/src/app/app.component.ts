import {Component, OnInit} from '@angular/core';
import {HTTPService} from './http.service';
import {MoviesService} from './movies.service';
import {AuthenticationService} from './authentication.service';

@Component({
    selector: 'movie-picker-root',
    templateUrl: './app.component.html',
    styleUrls: ['./app.component.less'],
})
export class AppComponent implements OnInit {
    constructor(private http: HTTPService, private moviesService: MoviesService, private auth: AuthenticationService) {}

    ngOnInit() {
        const {login, hash} = this.auth.getAuthInfo();

        if (login && hash) {
          this.moviesService.fetchUserLists(login, hash);
        }

        this.http.getLists().subscribe(this.moviesService.changeLists);
        this.moviesService.selectList();
    }
}
