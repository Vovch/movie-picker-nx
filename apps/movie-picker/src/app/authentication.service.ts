import {Injectable} from '@angular/core';
import {HTTPService} from './http.service';
import {BehaviorSubject, tap} from 'rxjs';
import { IMovie, IUserMovieList } from '@movie-picker/api-interfaces';

@Injectable({
  providedIn: 'root',
})
export class AuthenticationService {
  private _isAuthenticated = new BehaviorSubject<boolean>(false);
  private _login = localStorage.getItem('login');
  private _hash = localStorage.getItem('hash');

  isAuthenticated$ = this._isAuthenticated.asObservable();

  constructor(private http: HTTPService) {
    this._isAuthenticated.next(Boolean(this._login && this._hash));
  }

  async generateHash(login: string, password: string) {
    const encoder = new TextEncoder();

    const hash = await crypto.subtle.digest('SHA-256', encoder.encode(login.toLowerCase() + password));
    const hashHex = [...new Uint8Array(hash)].map((byte) => byte.toString(16).padStart(2, '0')).join('');

    return hashHex;
  }

  getAuthInfo() {
    return {login: this._login, hash: this._hash};
  }

  private storeAuthInfo(login: string, hash: string) {
    this._login = login;
    this._hash = hash;

    localStorage.setItem('login', login.toLowerCase());
    localStorage.setItem('hash', hash);

    this._isAuthenticated.next(true);
  }

  async login(login: string, password: string) {
    const hash = await this.generateHash(login, password);

    const loginSubject = this.http.login(login, hash);

    return loginSubject.pipe(tap(() => {
      this.storeAuthInfo(login, hash);
    }));
  }

  async register(login: string, password: string, captcha: string) {
    const hash = await this.generateHash(login, password);

    const registrationSubject = this.http.register(login, hash, captcha);

    return registrationSubject.pipe(tap(() => {
      this.storeAuthInfo(login, hash);
      this.login(login, hash);
    }));
  }

  logOut() {
    this._login = null;
    this._hash = null;

    localStorage.removeItem('login');
    localStorage.removeItem('hash');

    this._isAuthenticated.next(false);
  }
}
