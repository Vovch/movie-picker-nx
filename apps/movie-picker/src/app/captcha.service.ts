import {Injectable} from '@angular/core';

@Injectable({
    providedIn: 'root',
})
export class CaptchaService {
    #loginCaptchaId?: number;
    #registrationCaptchaId?: number;

    set loginCaptchaId(id: number | undefined) {
        this.#loginCaptchaId = id;
    }

    get loginCaptchaId() {
        return this.#loginCaptchaId;
    }

    set registrationCaptchaId(id: number | undefined) {
        this.#registrationCaptchaId = id;
    }

    get registrationCaptchaId() {
        return this.#registrationCaptchaId;
    }
}
