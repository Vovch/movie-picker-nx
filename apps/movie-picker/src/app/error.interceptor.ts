import {Injectable} from '@angular/core';
import {HttpRequest, HttpHandler, HttpEvent, HttpInterceptor, HttpErrorResponse} from '@angular/common/http';
import {catchError, Observable, throwError} from 'rxjs';
import {Store} from '@ngrx/store';
import {showErrorMessage} from './error/error.actions';

@Injectable()
export class ErrorInterceptor implements HttpInterceptor {
    constructor(private store: Store) {}

  intercept(request: HttpRequest<unknown>, next: HttpHandler): Observable<HttpEvent<unknown>> {
    return next.handle(request).pipe(
        catchError((error: HttpErrorResponse) => {
            let errorMsg = '';
            if (error.error instanceof ErrorEvent) {
                console.log('This is client side error');
                errorMsg = `Error: ${error.error.message}`;
            } else {
                console.log('This is server side error');
                errorMsg = `Error Code: ${error.status},  Message: ${error.message}`;
            }

            this.store.dispatch(showErrorMessage({message: 'Something went wrong.'}))
            console.log(errorMsg);
            return throwError(() => errorMsg);
        }),
    );
  }
}
