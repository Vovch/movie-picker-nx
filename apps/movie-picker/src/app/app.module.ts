import { NgModule, isDevMode } from '@angular/core';
import { BrowserModule } from '@angular/platform-browser';

import { AppRoutingModule } from './app-routing.module';
import { AppComponent } from './app.component';
import { HeaderComponent } from './header/header.component';
import { MovieListPageComponent } from './movie-list-page/movie-list-page.component';
import { MovieListComponent } from './movie-list/movie-list.component';
import { HTTP_INTERCEPTORS, HttpClientModule } from '@angular/common/http';
import { SvgModule } from './svg/svg.module';
import { SharedModule } from './shared/shared.module';
import { ReactiveFormsModule } from '@angular/forms';
import { AuthenticationModalComponent } from './authentication-modal/authentication-modal.component';
import { MovieModalComponent } from './movie-modal/movie-modal.component';
import { RegistrationModalComponent } from './registration-modal/registration-modal.component';
import { ErrorInterceptor } from './error.interceptor';
import { StoreModule } from '@ngrx/store';
import { ErrorComponent } from './error/error.component';
import { errorReducer } from './error/error.reducer';
import { StoreDevtoolsModule } from '@ngrx/store-devtools';
import { environment } from '../environments/environment';
import { ServiceWorkerModule } from '@angular/service-worker';

@NgModule({
  declarations: [
    AppComponent,
    HeaderComponent,
    MovieListPageComponent,
    MovieListComponent,
    AuthenticationModalComponent,
    MovieModalComponent,
    RegistrationModalComponent,
    ErrorComponent,
  ],
  imports: [
    BrowserModule,
    AppRoutingModule,
    HttpClientModule,
    SvgModule,
    SharedModule,
    ReactiveFormsModule,
    StoreModule.forRoot({ errors: errorReducer }, {}),
    StoreDevtoolsModule.instrument({
      maxAge: 25,
      logOnly: environment.production,
    }),
    StoreDevtoolsModule.instrument({
      maxAge: 25, // Retains last 25 states
      logOnly: environment.production, // Restrict extension to log-only mode
      autoPause: true, // Pauses recording actions and state changes when the extension window is not open
    }),
    ServiceWorkerModule.register('ngsw-worker.js', {
      enabled: !isDevMode(),
      // Register the ServiceWorker as soon as the application is stable
      // or after 30 seconds (whichever comes first).
      registrationStrategy: 'registerWhenStable:30000'
    }),
  ],
  providers: [
    {
      provide: HTTP_INTERCEPTORS,
      useClass: ErrorInterceptor,
      multi: true,
    },
  ],
  bootstrap: [AppComponent],
})
export class AppModule {}
