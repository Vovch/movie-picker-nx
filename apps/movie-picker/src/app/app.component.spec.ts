import { ComponentFixture, TestBed } from '@angular/core/testing';
import { RouterTestingModule } from '@angular/router/testing';
import { AppComponent } from './app.component';
import {HTTPService} from "./http.service";
import {of} from "rxjs";
import {NO_ERRORS_SCHEMA} from "@angular/core";
import { MoviesService } from './movies.service';
import { AuthenticationService } from './authentication.service';
import { IMovie } from '@movie-picker/api-interfaces';

describe('AppComponent', () => {
  let httpService: HTTPService;
  let auth: AuthenticationService;
  let moviesService: MoviesService;
  let fixture: ComponentFixture<AppComponent>;
  let appComponent: AppComponent;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [
        RouterTestingModule
      ],
      declarations: [
        AppComponent
      ],
      providers: [{provide: HTTPService, useValue: { getMovies: jest.fn().mockReturnValue(of([])) }}],
      schemas: [NO_ERRORS_SCHEMA],
    }).compileComponents();

    httpService = TestBed.inject(HTTPService);
    auth = TestBed.inject(AuthenticationService);
    moviesService = TestBed.inject(MoviesService);
    fixture = TestBed.createComponent(AppComponent);
    appComponent = fixture.componentInstance;
  });

  it('should create the app', () => {
    const fixture = TestBed.createComponent(AppComponent);
    const app = fixture.componentInstance;
    expect(app).toBeTruthy();
  });

  it('should render title', () => {
    const fixture = TestBed.createComponent(AppComponent);
    fixture.detectChanges();
    const compiled = fixture.nativeElement as HTMLElement;
    expect(compiled).toMatchSnapshot();
  });

  it('should fetch user lists when user is authenticated', () => {
    jest.spyOn(auth, 'getAuthInfo').mockReturnValue({ login: 'test', hash: 'test' });
    jest.spyOn(moviesService, 'fetchUserLists').mockImplementation();

    appComponent.ngOnInit();

    expect(moviesService.fetchUserLists).toHaveBeenCalledWith('test', 'test');
  });

  it('should not fetch user lists when user is not authenticated', () => {
    jest.spyOn(auth, 'getAuthInfo').mockReturnValue({ login: null, hash: null });
    jest.spyOn(moviesService, 'fetchUserLists').mockImplementation();

    appComponent.ngOnInit();

    expect(moviesService.fetchUserLists).not.toHaveBeenCalled();
  });

  it('should change movies list when HTTP request is successful', () => {
    const mockMovies: IMovie[] = [
      {
        id: 1,
        name: 'Movie 1',
        originalName: 'Original Movie 1',
        director: 'Director 1',
        yearProduced: '2001',
        yearAdded: '2002'
      },
      {
        id: 2,
        name: 'Movie 2',
        originalName: 'Original Movie 2',
        director: 'Director 2',
        yearProduced: '2003',
        yearAdded: '2004'
      },
      {
        id: 3,
        name: 'Movie 3',
        originalName: 'Original Movie 3',
        director: 'Director 3',
        yearProduced: '2005',
        yearAdded: '2006'
      },
    ];

    jest.spyOn(httpService, 'getMovies').mockReturnValue(of({ list: mockMovies, listId: '', name: '' }));
    jest.spyOn(moviesService, 'changeMoviesList').mockImplementation();

    appComponent.ngOnInit();

    expect(moviesService.changeMoviesList).toHaveBeenCalledWith({ list: mockMovies, listId: '', name: '' });
  });
});
