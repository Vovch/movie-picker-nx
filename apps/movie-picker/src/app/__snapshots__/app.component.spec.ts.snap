// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppComponent should render title 1`] = `
<div
  id="root1"
>
  <movie-picker-header />
  <section
    class="router-outlet"
  >
    <router-outlet />
    
    <movie-picker-authentication-modal />
    <movie-picker-error />
  </section>
</div>
`;
